hwy/ops/ppc_vsx-inl.h

```diff
@@ -0,0 +1,4168 @@
+template <typename T, size_t N>
+HWY_API Vec128<T, N> Xor3(Vec128<T, N> x1, Vec128<T, N> x2, Vec128<T, N> x3) {
+  return Xor(x1, Xor(x2, x3));
+}
+
+// ------------------------------ Or3
+template <typename T, size_t N>
+HWY_API Vec128<T, N> Or3(Vec128<T, N> o1, Vec128<T, N> o2, Vec128<T, N> o3) {
+  return Or(o1, Or(o2, o3));
+}
+
+// ------------------------------ OrAnd
+template <typename T, size_t N>
+HWY_API Vec128<T, N> OrAnd(Vec128<T, N> o, Vec128<T, N> a1, Vec128<T, N> a2) {
+  return Or(o, And(a1, a2));
+}
+
+// ------------------------------ IfVecThenElse
+template <typename T, size_t N>
+HWY_API Vec128<T, N> IfVecThenElse(Vec128<T, N> mask, Vec128<T, N> yes,
+                                   Vec128<T, N> no) {
+  return Vec128<T, N>{(typename detail::Raw128<T>::type)vec_sel(
+    (typename detail::Raw128<MakeUnsigned<T>>::type)no.raw,
+    (typename detail::Raw128<MakeUnsigned<T>>::type)yes.raw,
+    (typename detail::Raw128<MakeUnsigned<T>>::type)mask.raw)};
+}
+
+// ------------------------------ Operator overloads (internal-only if float)
+
+template <typename T, size_t N>
+HWY_API Vec128<T, N> operator&(Vec128<T, N> a, Vec128<T, N> b) {
+  return And(a, b);
+}
+
+template <typename T, size_t N>
+HWY_API Vec128<T, N> operator|(Vec128<T, N> a, Vec128<T, N> b) {
+  return Or(a, b);
+}
+
+template <typename T, size_t N>
+HWY_API Vec128<T, N> operator^(Vec128<T, N> a, Vec128<T, N> b) {
+  return Xor(a, b);
+}
+
+// ================================================== SIGN
+
+// ------------------------------ Neg
+
+template <typename T, size_t N, HWY_IF_NOT_SPECIAL_FLOAT(T)>
+HWY_INLINE Vec128<T, N> Neg(Vec128<T, N> v) {
+  return Vec128<T, N>{vec_neg(v.raw)};
+}
+
+// ------------------------------ Abs
+
+// Returns absolute value, except that LimitsMin() maps to LimitsMax() + 1.
+template <class T, size_t N, HWY_IF_NOT_SPECIAL_FLOAT(T)>
+HWY_API Vec128<T, N> Abs(Vec128<T, N> v) {
+  return Vec128<T, N>{vec_abs(v.raw)};
+}
+
+// ------------------------------ CopySign
+
+template <typename T, size_t N>
+HWY_API Vec128<T, N> CopySign(Vec128<T, N> magn, Vec128<T, N> sign) {
+  static_assert(IsFloat<T>(), "Only makes sense for floating-point");
+  return Vec128<T, N>{vec_cpsgn(sign.raw, magn.raw)};
+}
+
+template <typename T, size_t N>
+HWY_API Vec128<T, N> CopySignToAbs(Vec128<T, N> abs, Vec128<T, N> sign) {
+  // PPC8 can also handle abs < 0, so no extra action needed.
+  return CopySign(abs, sign);
+}
+
+// ================================================== MEMORY (1)
+
+// ------------------------------ Load
+
+template <class D, HWY_IF_V_SIZE_D(D, 16), typename T = TFromD<D>>
+HWY_API Vec128<T> Load(D /* tag */, const T* HWY_RESTRICT aligned) {
+  using LoadRawVectType =
+    typename detail::Raw128<T>::AlignedLoadStoreRawVectType;
+  using ResultRawVectType =
+    typename detail::Raw128<T>::type;
+  return Vec128<T>{(ResultRawVectType)
+    *reinterpret_cast<const LoadRawVectType*>(aligned)};
```

Comment:
Is the type punning safe also for clang? Should we instead use memcpy/CopyBytes?

Reply:
The reason why a type punning cast is used (which is safe in the code above on GCC/Clang as the pointer is being cast to a vector type that has the ```__may_alias__``` attribute) is that older versions of GCC don't optimize the memcpy of a full 16-byte vector load down to a single ```lvx``` instruction but optimize the above type-punning vector load down to a single ```lvx``` instruction.

Newer versions of GCC such as GCC 11 and GCC 12 will optimize CopyBytes down to a single ```lvx``` instruction if ```__builtin_assume_aligned(aligned, 16)``` is passed into ```__builtin_memcpy```.

Here is the alternative implementation of a full 16-byte vector load using CopyBytes and HWY_ASSUME_ALIGNED:
```
template <class D, HWY_IF_V_SIZE_D(D, 16), typename T = TFromD<D>>
HWY_API Vec128<T> Load(D /* tag */, const T* HWY_RESTRICT aligned) {
  typename detail::Raw128<T>::type result;
  CopyBytes<sizeof(detail::Raw128<T>::type)>(
    HWY_ASSUME_ALIGNED(aligned, 16), &result);
  return result;
}
```

Note that HWY_ASSUME_ALIGNED is needed above for GCC/Clang to treat the above load as an aligned load.

Reply:
Got it, thanks for explaining. Seems safer to keep the current type punning, then. But I'll add a comment to make clear why we are doing this.